doc/interactive_lua.de.man
.\" Process this file with
.\" groff -man -Tutf8 interactive_lua.en.man
.\"
.
.\" Macros .TQ .EX .EE taken from groff an-ext.tmac
.\" Copyright (C) 2007, 2009 Free Software Foundation, Inc.
.\" You may freely use, modify and/or distribute this file.
.
.\" Continuation line for .TP header.
.de TQ
.  br
.  ns
.  TP \\$1\" no doublequotes around argument!
..
.
.\" Start example.
.de EX
.  nr mE \\n(.f
.  nf
.  nh
.  ft CW
..
.
.
.\" End example.
.de EE
.  ft \\n(mE
.  fi
.  hy \\n(HY
..
.
.TH "interactive_lua.lua" 1 2012-07-24 AKFAvatar
.nh
.
.SH NAME
interactive_lua.lua \- Lua interaktiv mit AKFAvatar verwenden
.
.SH SYNTAX
.BI interactive_lua.lua
.
.SH BESCHREIBUNG
Lua interaktiv mit AKFAvatar verwenden.
.PP
Man kann es zum Beispiel als Taschenrechner verwenden.
Wenn man eine Zeile mit einem Gleichheitszeichen (\[Bq]=\[lq]) einleitet,
wird das Ergebnis des folgenden Ausdrucks angezeigt.
Zum Beisiel, wenn man \[Bq]=10/4\[lq] eingibt, gibt er \[Bq]2,5\[lq] aus.
.PP
Zu beachten ist, dass das Dezimalzeichen, welches angezeigt wird, abh\[:a]ngig von
den Sprach-Einstellungen ist.
Wenn man jedoch einen Dezimalbruch eingeben will, muss man die amerikanische
Notation mit einem Dezimal-Punkt verwenden!
.PP
Man kann seine eigenen Variablen definieren.
Zum Beispiel ein Megabyte ist 2^20 Bytes.
Man kann also \[Bq]MB=2^20\[lq] eingeben, und dann kann man das in eigenen
Berechnungen verwenden. Wenn man wissen will, wieviel 5 Megabytes in Bytes
sind, kann man \[Bq]=5*MB\[lq] eingeben, das Ergebnis ist dann
\[Bq]5242880\[lq].
Man mus auch darauf achten, dass bei Lua die Gro\[ss]-/Klein-Schreibung wichtig
ist.
Das bedeutet, dass \[Bq]MB\[lq] und \[Bq]mb\[lq] unterschiedliche
Variablen sind.
.PP
Die M\[:o]glichkeit Zeilen mit einem Gleichheitszeichen einzuleiten ist eine
besondere Eigenschaft des interaktiven Programmes.
Es kann nicht in Skripten verwendet werden.
.PP
Die normale Eingabeaufforderung ist ein \[Bq]>\[lq].
Wenn man jedoch eine unvollst\[:a]ndinge Zeile eingibt, bekommt man eine andere
Eingabeaufforderung, n\[:a]mlich \[Bq]>>\[lq].
Dann kann man seine Eingabe fortsetzen.
Wenn man versehentlich bei dieser Eingabeaufforderung landet, kann man
einfach <Enter> dr\[:u]cken.
Das verwirft die vorherige Eingabe.
.PP
Man kann auch seine eigenen Funktionen definieren.
Zum Beispiel:
.EX
  > function Kreisflaeche(radius)
  >> return radius^2 * math.pi
  >> end
  > =3 * Kreisflaeche(5)
  235.61944901923
.EE
.PP
Man kann jeden Lua-Befehl verwenden, einschlie\[ss]lich AKFAvatar-Befehlen.
Zum Beispiel kann man dem Avatar einen Namen geben:
.EX
  > avt.set_avatar_name("Hugo")
.EE
.PP
Oder, wenn man \[:u]berhaupt keinen Avatar haben will:
.EX
  > avt.avatar_image("none")
.EE
.PP
Um einen weiblichen Avatar zu bekommen:
.EX
  > avt.avatar_image_file(avt.search("female_user.xpm"))
.EE
.PP
Man kann so viel mehr damit machen...
Einfach mal rumprobieren...
.PP


NEWS
* AKFAvatar-testing

 - optimized printing of text (line buffered)
 - audio-data with same specs as former one is played with less latency
 - improved support for 24 and 32 bit audio
 - new file Bücher.lua for reading German books from my server (needs curl)

  C-API changes:
    - new macro: AVT_KEY_F

* AKFAvatar 0.24.3

  - audio: big audio files are read directly from file while playing
    (if possible via mmap)
  - audio: AU and WAV data can be read from non-seekable streams
    (loaded into memory)
  - fix: Pascal sound support fixed
  - the control characters SO/SI switch underlined mode on/off;
    the control characters DC1/DC2 switch bold mode on/off
  - configure: new option --with-sdl2
  - SDL2: support for SDL2 with automatic resizing of the graphics
  - SDL2: support for pasting text from clipboard (Alt+V, middle mouse button)

  Lua-API changes:
    - new functions: avt.frequency
    - module "akfavatar-term": fix APC interface

* AKFAvatar 0.24.2

  - lua/clock.lua: resizing works (with some delay)
  - lua/four_in_a_row.lua: mouse support / resizing
  - lua/four_in_a_row.lua: play against computer

  Lua-API changes:
    - Module "akfavatar-graphic":
      - new functions: graphic.set_pointer_buttons_key,
        graphic.set_pointer_motion_key, graphic.get_pointer_position,
        graphic.set_resize_key

* AKFAvatar 0.24.1

  - fix: lua/interactive_lua.lua: fix handling of incomplete lines
  - Lua module "akfavatar-graphic" included in executable
  - text cursor modified
  - lua-akfavatar: GNU .info files can be shown as plain text (no hyperlinks)
  - lua/hexviewer.lua: show more characters on the right side
  - lua/multiplication.lua: clear board (no scrolling)
  - configure accepts the option --with-linuxfb
    which then builds for the Linux framebuffer without the SDL
    (no mouse and no audio support yet)

  C-API changes:
    - new functions: avt_detect_utf8

  Lua-API changes:
    - new functions: avt.toutf8, avt.detect_utf8, avt.utf8codepoints

* AKFAvatar 0.24.0

  - fix: terminal doesn't use so much CPU time anymore
  - fix: window resizing for the pager works again
  - fix: switching fullscreen mode while avatar moves in or out
  - fix: avt_credits sometimes got the end wrong
  - fix: audioplayer.lua accepts PLS playlists again
  - fix: Pascal binding works again, extended
  - paging through menus is animated
  - lua/audioplayer.lua: improved playlist support
  - lua/multiplication.lua: more authentic

  C-API changes:
    - avt_char is unsigned int now, avt_choice adapted
      should be compatible with char32_t in C11
    - functions with *_mb* replaced with functions *_char*
    - new functions: avt_darker, avt_brighter, avt_brightness
    - avtadons.h also uses the prefix avt_ for everything

  Lua-API changes:
    - The default encoding is now ASCII
      to encourage always using avt.encoding()
    - avt.set_directory or avt.chdir return true on success
      or nil and an error message on error

  Pascal-API changes:
    - variable CheckEsc removed
    - new procedures/functions: Menu, ReadKeyCode, reserve_single_keys


* AKFAvatar 0.23.1

  - fixed vulnerabilities
  - avoid recursive calls of the bell
  - split source code for better maintainablitity
  - name sign with 3d effect

  C-API changes:
    - avt_set_title: no automatic charset conversion
    - avt_put_char: support for UTF-16 surrogate characters
    - new functions: avt_wide_copyright, avt_wide_license, avt_wide_version,
      avt_menu, avt_palette_size
    - avtaddons:
      - avta_file_selection and avta_filter have a new parameter "data",
        which is passed along

  Lua-API changes:
    - Module "akfavatar-graphic"
      - new methods: gr:border3d

* AKFAvatar 0.23.0

  The code was made less dependend on SDL. There is a new experimantal
  backend for a Linux framebuffer device, but without mouse support or audio
  (of course you can call an external audio-player in your scripts).

  - bugfix: fix for XPM images with more than 256 colors
  - bugfix: lua-akfavatar also resets the charset
  - bugfix/change: lua-akfavatar option -l sets a global variable
  - avt_update() is much faster
  - support for combining characters (Unicode)
  - The pager supports numbers (number-pad when NumLock is on)
  - lua/interactive_lua.lua supports scrolling back in a history list
  - regression: the navigation bar cannot handle the pause key anymore
  - key codes have been changed not to conflict with the "corporate zone"

  C-API changes:
    - all deprecated functions removed
    - also removed: avt_register_keyhandler, avt_register_mousehandler
    - avt_update() is much faster
    - avt_get_key replaces avt_key
    - modified: avt_show_raw_image, avt_put_raw_image_file,
      avt_put_raw_image_stream, avt_put_raw_image_data, avt_put_raw_image_xpm
      don't have the bpp parameter anymore, only 4 Bytes per pixel accepted
    - new functions: avt_key_pressed, avt_push_key, avt_clear_keys,
      avt_set_audio_end_key, avt_combining, avt_bell_function
    - new experimental functions: avt_set_pointer_buttons_key,
      avt_set_pointer_motion_key, avt_get_pointer_position,
      avt_char avt_input, avt_input_mb
    - new macro: avt_elapsed

  Lua-API changes:
    - the function avt.ask was extended (yet undocumented)
    - new functions: avt.combining, avt.key_pressed, avt.clear_keys,
      avt.push_key, avt.set_audio_end_key, avt.alert, avt.getcwd
    - new table: avt.key with key codes


* AKFAvatar 0.22.1

  There are a lot of changes in the APIs. Also the ABI was changed.

  - Window is resizable again (content gets centered)
  - The image can be a separated banner as header or footer
  - Uses Lua-5.2 now, no longer compatible with 5.1
  - Lua modules don't define globals anymore (scripts have to be adapted!)
  - Data files are put in a separate directory and Lua has a search function
    for it (scripts have to be adapted!)
  - the function for translations is integrated in Lua-AKFAvatar

lua/interactive_lua.lua
#!/usr/bin/env lua-akfavatar

-- based on code from the Lua wiki - author unknown

avt = require "lua-akfavatar"

print = avt.print --> redefine the print command
say = print --> an alias that fits better

local history = {}

local function interactive (cmd)

  local function input (prompt)
    local pos = 0

    avt.save_position ()
    local line, ch = avt.input (prompt, nil, -1, 1)

    while ch == avt.key.up or ch == avt.key.down do
      if ch == avt.key.up then
        if history[pos+1] then
          pos = pos + 1
        end
      elseif ch == avt.key.down then
        if history[pos] then
          pos = pos - 1
        end
      end
      avt.restore_position ()
      line, ch = avt.input (prompt, history[pos], -1, 1)
    end

    if line ~= "" then table.insert (history, 1, line) end

    return line
  end

  local function error_message (...)
    avt.bell ()
    avt.set_text_color ("dark red")
    avt.print (...)
    avt.normal_text ()
  end

  local function show (success, ...)
    if select ("#", ...) > 0 then
      if success
        then avt.print (...)
        else error_message (...)
      end --> if success
    end --> if select
  end --> function show

  if avt.where_x () > 1 then avt.newline () end

  if not cmd then --> first line
    -- replace "=" at the beginning with "return "
    cmd = string.gsub (input ("> "), "^=", "return ", 1)
  else
    local cmd2 = input (">> ")
    if cmd2 == "" then return "" end --> empty line cancels command
    cmd = cmd .. " " .. cmd2
  end

  local func, err = load (cmd)

  if func then show (pcall(func))
    else --> error
      -- <eof> at the end means the command is incomplete
      if (string.find(err, "<eof>"))
        then return interactive (cmd)
        else error_message (err)
      end
  end

  return cmd
end

avt.encoding("UTF-8")
avt.title("Lua-AKFAvatar")
avt.start()
avt.start_audio()
avt.avatar_image_file(avt.search("computer.xpm") or "default")
avt.set_avatar_mode("footer")

avt.underlined(true)
avt.say(_VERSION, " / AKFAvatar ", avt.version (), "\n\n")
avt.underlined(false)

while true do interactive () end
